### ISSUE TITLE
Using torch.compile, batch training step time takes long to converge when adding a LR Scheduler

### ISSUE THREAD

**Issue by visionscaper** [pytorch/pytorch#120934]
### 🐛 Describe the bug

This bug report is created based on this thread on the PyTorch Forums:
["Torch._dynamo hit config.cache_size_limit when adding LR scheduler"](https://discuss.pytorch.org/t/torch-dynamo-hit-config-cache-size-limit-when-adding-lr-scheduler/197738)

It is observed that the duration of a batch training step, compiled with `torch.compile`, takes longer to converge when adding an LR scheduler to the end of the step function. For some reason it seems that more compilations are cached.

Tagging @msaroufim on request.

### Error logs

There are no errors, but warnings that signal that there is an issue, also see the PyTorch forum thread as mentioned above:

After the first 12 steps:

> torch._dynamo.convert_frame: [WARNING] torch._dynamo hit config.cache_size_limit (8)
 torch._dynamo.convert_frame: [WARNING]    function: 'step' (/root/.virtualenvs/mlpug/lib/python3.10/site-packages/torch/optim/adamw.py:151)
> torch._dynamo.convert_frame: [WARNING]    **last reason: L['self'].param_groups[0]['lr'] == 4.8361308593749987e-05     # has_complex = self._init_group(  # optim/adamw.py:176 in step**
> torch._dynamo.convert_frame: [WARNING] To log all recompilation reasons, use TORCH_LOGS="recompiles".
> torch._dynamo.convert_frame: [WARNING] To diagnose recompilation issues, see https://pytorch.org/docs/master/compile/troubleshooting.html.

After step 16:

> ...
torch._dynamo.convert_frame: [WARNING]    function: 'adamw' (/root/.virtualenvs/mlpug/lib/python3.10/site-packages/torch/optim/adamw.py:279)
torch._dynamo.convert_frame: [WARNING]    **last reason: L['lr'] == 3.208388020496905e-05                              # if foreach and isinstance(lr, Tensor) and not capturable:  # optim/adamw.py:320 in adamw**
...

After step 20:

> ...
torch._dynamo.convert_frame: [WARNING]    function: '_multi_tensor_adamw' (/root/.virtualenvs/mlpug/lib/python3.10/site-packages/torch/optim/adamw.py:479)
 torch._dynamo.convert_frame: [WARNING]    **last reason: L['lr'] == 2.6132520951198568e-05                             # if isinstance(lr, Tensor) and not capturable:  # optim/adamw.py:503 in _multi_tensor_adamw**
...

### Versions

```
  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100 22068  100 22068    0     0  54684      0 --:--:-- --:--:-- --:--:-- 54759
Collecting environment information...
PyTorch version: 2.3.0.dev20240228+cu121
Is debug build: False
CUDA used to build PyTorch: 12.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 22.04.3 LTS (x86_64)
GCC version: (Ubuntu 11.4.0-1ubuntu1~22.04) 11.4.0
Clang version: Could not collect
CMake version: version 3.26.4
Libc version: glibc-2.35

Python version: 3.10.13 (main, Sep 11 2023, 13:44:35) [GCC 11.2.0] (64-bit runtime)
Python platform: Linux-5.15.0-87-generic-x86_64-with-glibc2.35
Is CUDA available: True
CUDA runtime version: Could not collect
CUDA_MODULE_LOADING set to: LAZY
GPU models and configuration: 
GPU 0: NVIDIA RTX A6000
GPU 1: NVIDIA RTX A6000
GPU 2: NVIDIA RTX A6000
GPU 3: NVIDIA RTX A6000

Nvidia driver version: 535.113.01
cuDNN version: Could not collect
HIP runtime version: N/A
MIOpen runtime version: N/A
Is XNNPACK available: True

CPU:
Architecture:                       x86_64
CPU op-mode(s):                     32-bit, 64-bit
Address sizes:                      46 bits physical, 48 bits virtual
Byte Order:                         Little Endian
CPU(s):                             48
On-line CPU(s) list:                0-47
Vendor ID:                          GenuineIntel
Model name:                         Intel(R) Xeon(R) Gold 6248R CPU @ 3.00GHz
CPU family:                         6
Model:                              85
Thread(s) per core:                 1
Core(s) per socket:                 24
Socket(s):                          2
Stepping:                           7
BogoMIPS:                           5986.98
Flags:                              fpu vme de pse tsc msr pae mce cx8 apic sep mtrr pge mca cmov pat pse36 clflush mmx fxsr sse sse2 ss ht syscall nx pdpe1gb rdtscp lm constant_tsc arch_perfmon rep_good nopl xtopology cpuid tsc_known_freq pni pclmulqdq vmx ssse3 fma cx16 pdcm pcid sse4_1 sse4_2 x2apic movbe popcnt tsc_deadline_timer aes xsave avx f16c rdrand hypervisor lahf_lm abm 3dnowprefetch cpuid_fault invpcid_single ssbd ibrs ibpb stibp ibrs_enhanced tpr_shadow vnmi flexpriority ept vpid ept_ad fsgsbase tsc_adjust bmi1 avx2 smep bmi2 erms invpcid mpx avx512f avx512dq rdseed adx smap clflushopt clwb avx512cd avx512bw avx512vl xsaveopt xsavec xgetbv1 xsaves arat umip pku ospke avx512_vnni md_clear flush_l1d arch_capabilities
Virtualization:                     VT-x
Hypervisor vendor:                  KVM
Virtualization type:                full
L1d cache:                          1.5 MiB (48 instances)
L1i cache:                          1.5 MiB (48 instances)
L2 cache:                           192 MiB (48 instances)
L3 cache:                           32 MiB (2 instances)
NUMA node(s):                       1
NUMA node0 CPU(s):                  0-47
Vulnerability Gather data sampling: Unknown: Dependent on hypervisor status
Vulnerability Itlb multihit:        Not affected
Vulnerability L1tf:                 Not affected
Vulnerability Mds:                  Not affected
Vulnerability Meltdown:             Not affected
Vulnerability Mmio stale data:      Mitigation; Clear CPU buffers; SMT Host state unknown
Vulnerability Retbleed:             Mitigation; Enhanced IBRS
Vulnerability Spec rstack overflow: Not affected
Vulnerability Spec store bypass:    Mitigation; Speculative Store Bypass disabled via prctl and seccomp
Vulnerability Spectre v1:           Mitigation; usercopy/swapgs barriers and __user pointer sanitization
Vulnerability Spectre v2:           Mitigation; Enhanced IBRS, IBPB conditional, RSB filling, PBRSB-eIBRS SW sequence
Vulnerability Srbds:                Not affected
Vulnerability Tsx async abort:      Mitigation; TSX disabled

Versions of relevant libraries:
[pip3] numpy==1.26.4
[pip3] pytorch-triton==3.0.0+901819d2b6
[pip3] torch==2.3.0.dev20240228+cu121
[pip3] torchaudio==2.2.0.dev20240228+cu121
[pip3] torchvision==0.18.0.dev20240228+cu121
[pip3] triton==2.2.0
[conda] blas                      1.0                         mkl  
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] libjpeg-turbo             2.0.0                h9bf148f_0    pytorch
[conda] mkl                       2023.1.0         h213fc3f_46344  
[conda] mkl-service               2.4.0           py310h5eee18b_1  
[conda] mkl_fft                   1.3.8           py310h5eee18b_0  
[conda] mkl_random                1.2.4           py310hdb19cb5_0  
[conda] numpy                     1.26.3          py310h5f9d8c6_0  
[conda] numpy-base                1.26.3          py310hb5e798b_0  
[conda] optree                    0.10.0                   pypi_0    pypi
[conda] pytorch                   2.2.1           py3.10_cuda12.1_cudnn8.9.2_0    pytorch
[conda] pytorch-cuda              12.1                 ha16c6d3_5    pytorch
[conda] pytorch-mutex             1.0                        cuda    pytorch
[conda] torchaudio                2.2.1               py310_cu121    pytorch
[conda] torchelastic              0.2.2                    pypi_0    pypi
[conda] torchtriton               2.2.0                     py310    pytorch
[conda] torchvision               0.17.1              py310_cu121    pytorch
```

cc @vincentqb @jbschlosser @albanD @janeyx99 @crcrpar @ezyang @msaroufim @bdhirsh @anijain2305 @zou3519 @chauhang @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @chenyang78 @kadeng @aakhundov

**Comment by desertfire** [comment]
cc @mlazos

**Comment by ezyang** [comment]
cc @janeyx99

**Comment by janeyx99** [comment]
This looks to be the same issue as https://github.com/pytorch/pytorch/issues/107076

**Comment by visionscaper** [comment]
@janeyx99 based on your comment I was able to resolve the problem. Instead of [converting the learning rate to a tensor in the optimizer](https://github.com/pytorch/pytorch/issues/107076#issuecomment-1956762299), I wrapped the output of my LR scheduling function, to be used with `LambdaLR`, in a tensor. In this way I didn't have to touch any PyTorch code.

So I changed my code from this:
```
scheduler = LambdaLR(optimizer, lr_scheduling_func)
```
To this
```
lr_scheduling_func = lambda iter: torch.tensor(lr_scheduling_func(iter), requires_grad=False)
scheduler = LambdaLR(optimizer, lr_scheduling_func)
```

Interestingly enough, the convergence *with* `scheduler.step()` is now faster than without!
These are my results (on a slightly different machine, but still 4x A6000):
```
-----------------------------------------------------------------------------------
| Epoch | Batch step | Batch training time     | Batch training time              |
|       |            | (NO LR scheduler)       | (Using scheduler.step() with fix)|
|       |            | (PyTorch Nightly+Cu121) | (PyTorch Nightly+Cu121)          |
-----------------------------------------------------------------------------------
|   0   |    10/976  |         16557ms         |             14942ms              |
|   0   |    20/976  |          8841ms         |              8034ms              |
|   0   |   300/976  |          1890ms         |              1830ms              |
|   0   |   900/976  |          1560ms         |              1538ms              |
|   2   |   430/976  |          1401ms         |              1401ms              |
-----------------------------------------------------------------------------------
```

I used the nightly of today:
```
% pip show torch
Name: torch
Version: 2.3.0.dev20240301+cu121
Summary: Tensors and Dynamic neural networks in Python with strong GPU acceleration
Home-page: https://pytorch.org/
Author: PyTorch Team
Author-email: packages@pytorch.org
License: BSD-3
Location: /root/.virtualenvs/mlpug/lib/python3.10/site-packages
Requires: filelock, fsspec, jinja2, networkx, nvidia-cublas-cu12, nvidia-cuda-cupti-cu12, nvidia-cuda-nvrtc-cu12, nvidia-cuda-runtime-cu12, nvidia-cudnn-cu12, nvidia-cufft-cu12, nvidia-curand-cu12, nvidia-cusolver-cu12, nvidia-cusparse-cu12, nvidia-nccl-cu12, nvidia-nvtx-cu12, pytorch-triton, sympy, typing-extensions
Required-by: torchaudio, torchvision
```

The fact that the convergence is faster with the scheduler+lr wrapped in tensor suggests to me that a lr tensor should be the default in optimizers. Could that be the case?

cc @msaroufim

**Comment by mlazos** [comment]
hi @visionscaper, this makes sense, the compiled optimizer has equal or greater precision than the non-compiled version due to fusions done in the compiler. Once composibility with the LR schedulers is clearer, it's possible we could change the default or do some wrapping under the hood, but we haven't gotten that far yet. This is a great datapoint though, thanks for figuring this out!

**Comment by mlazos** [comment]
Closing since this has been resolved by https://github.com/pytorch/pytorch/pull/125653
